# QBIO 401 Homework 8
## Hirad Hosseini, Fall 2023

Assignment #8
On Blackboard I have posted the data set wdbc.csv. In this assignment we are going to use
logistic regression to try to predict the column “diagnosis” (M for malignant or B for benign).

Each row is a different breast cancer patient. The first column is the ID_number of the patient
and this column does not concern us. The second column is diagnosis and this is the Y-variable
we are trying to predict. The next 30 columns are the numerical X-variables that we are going to
use to try to predict diagnosis. (Note: there is no missing data, and there are no categorical X-
variables so there is no need for “dummy” variables.) These 30 variables are computed from a
fine-scale image of the breast mass. The first 10 of these variables are mean values from the
cells. The next 10 of these variables are standard errors from the cells. And the last 10 of these
variables are the “worst” (average of the three largest cells).

Hint: See the code at the end of the notebook “regression2class1.ipynb” (in the “Regression 2
Classification 1 folder) in the “Logistic Regression” section

**Question 1: Randomly split the data into a training and test set (75% training and 25% set).
Standardize the X-variables. Fit the logistic regression model on the training set.
Calculate the fraction of times this model predicts the correct result on the test set.
Compute the confusion matrix on the test set.**

In [1]:
#Import all dependencies for logistic regression model and confusion matrix
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

In [2]:
df = pd.read_csv('wdbc.csv')
df

ID_number diagnosis  mean_radius  mean_texture  mean_perimeter  \
0       842302         M        17.99         10.38          122.80   
1       842517         M        20.57         17.77          132.90   
2     84300903         M        19.69         21.25          130.00   
3     84348301         M        11.42         20.38           77.58   
4     84358402         M        20.29         14.34          135.10   
..         ...       ...          ...           ...             ...   
564     926424         M        21.56         22.39          142.00   
565     926682         M        20.13         28.25          131.20   
566     926954         M        16.60         28.08          108.30   
567     927241         M        20.60         29.33          140.10   
568      92751         B         7.76         24.54           47.92   

     mean_area  mean_smoothness  mean_compactness  mean_concavity  \
0       1001.0          0.11840           0.27760         0.30010   
1       1326.0          0.08474           0.07864         0.08690   
2       1203.0          0.10960           0.15990         0.19740   
3        386.1          0.14250           0.28390         0.24140   
4       1297.0          0.10030           0.13280         0.19800   
..         ...              ...               ...             ...   
564     1479.0          0.11100           0.11590         0.24390   
565     1261.0          0.09780           0.10340         0.14400   
566      858.1          0.08455           0.10230         0.09251   
567     1265.0          0.11780           0.27700         0.35140   
568      181.0          0.05263           0.04362         0.00000   

     mean_concave_points  ...  worst_radius  worst_texture  worst_perimeter  \
0                0.14710  ...        25.380          17.33           184.60   
1                0.07017  ...        24.990          23.41           158.80   
2                0.12790  ...        23.570          25.53           152.50   
3                0.10520  ...        14.910          26.50            98.87   
4                0.10430  ...        22.540          16.67           152.20   
..                   ...  ...           ...            ...              ...   
564              0.13890  ...        25.450          26.40           166.10   
565              0.09791  ...        23.690          38.25           155.00   
566              0.05302  ...        18.980          34.12           126.70   
567              0.15200  ...        25.740          39.42           184.60   
568              0.00000  ...         9.456          30.37            59.16   

     worst_area  worst_smoothness  worst_compactness  worst_concavity  \
0        2019.0           0.16220            0.66560           0.7119   
1        1956.0           0.12380            0.18660           0.2416   
2        1709.0           0.14440            0.42450           0.4504   
3         567.7           0.20980            0.86630           0.6869   
4        1575.0           0.13740            0.20500           0.4000   
..          ...               ...                ...              ...   
564      2027.0           0.14100            0.21130           0.4107   
565      1731.0           0.11660            0.19220           0.3215   
566      1124.0           0.11390            0.30940           0.3403   
567      1821.0           0.16500            0.86810           0.9387   
568       268.6           0.08996            0.06444           0.0000   

     worst_concave_points  worst_symmetry  worst_fractal_dimension  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    ...             ...                      ...  
564           

In [3]:
y = df["diagnosis"] #feature to be predicted
temp_X = df.drop("diagnosis", axis=1) #all other features to be trained on 
X = pd.get_dummies(temp_X, drop_first=True)

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=100) #Random state of 100 
#for consistent output
scaler = StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

In [5]:
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

LogisticRegression()

In [6]:
log_reg.predict_proba(X_test[0:10]) #Predictions appear to be strongly in favor of one label

array([[2.93066065e-06, 9.99997069e-01],
       [6.72685921e-01, 3.27314079e-01],
       [6.19065613e-04, 9.99380934e-01],
       [9.99987712e-01, 1.22877755e-05],
       [9.97021098e-01, 2.97890240e-03],
       [9.74417970e-01, 2.55820301e-02],
       [3.33579966e-01, 6.66420034e-01],
       [1.03968055e-06, 9.99998960e-01],
       [9.50715099e-01, 4.92849012e-02],
       [9.25352228e-01, 7.46477717e-02]])

In [7]:
log_reg.predict(X_test[0:10]) 

array(['M', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B', 'B'], dtype=object)

In [8]:
y_test[0:10] #Predictions and ground truth mostly match for first 10 cases

400    M
225    B
321    M
173    B
506    B
380    B
197    M
260    M
40     M
160    B
Name: diagnosis, dtype: object

In [9]:
np.mean(y_test == log_reg.predict(X_test)) 

0.965034965034965

In [10]:
confusion_matrix(y_test, log_reg.predict(X_test))

array([[87,  0],
       [ 5, 51]])

This model predicts the correct diagnosis label approximatley 96.5% of the time for the test set. The confusion matrix demonstrates a high percent of true positives and true negatives, with a few false negatives. Overall, this logistic regression model is a good rudimentary start for predicting diagnoses. 

**Question 2: Next we are going to see how much our result in #1 depends on the initial random split
into a training and test set. Use a loop to repeat the following 100 times: randomly split
the data into a training and test split (75% training and 25% set), standardize the X-
variables, fit the logistic regression model on the training set, and calculate the fraction
of times this model predicts the correct result on the test set. (Note: you do not need to
compute the confusion matrix for this problem.) Once you have the 100 estimates of
the fraction of times the model predicts the correct result on the test set, report the
95% bootstrap confidence interval for this estimate (sort the 100 estimates, and report
the values at position 1 and 98).**

Hint: it is important that each split into training and test sets is different. If you don’t
specify “random_state” (see below), it will be different every time:

X_train, X_test, y_train, y_test = train_test_split(X, y,
test_size=0.25)

In [15]:
frac_correct_pred_estims = [] #contains all model accuracy scores for 100 trials
for trial in range(100): #conduct 100 trials for bootsrapping
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)
    scaler = StandardScaler().fit(X_train) #standardizes X variables
    X_train = scaler.transform(X_train)
    X_test = scaler.transform(X_test)
    log_reg = LogisticRegression()
    log_reg.fit(X_train, y_train) #fits logistic regression model to the data
    frac_correct_pred_estims.append(np.mean(y_test == log_reg.predict(X_test))) #obtains accuracy score for model
frac_correct_pred_estims.sort()
print(f"The 95% bootstrap confidence interval for the estimate fraction of correct predictions is [{frac_correct_pred_estims[1]}, {frac_correct_pred_estims[98]}]")

The 95% bootstrap confidence interval for the estimate fraction of correct predictions is [0.9440559440559441, 0.993006993006993]


Our obtained 95% confidence interval of approximately (0.944, 0.993) demonstrates that a training-test split of 75/25% is a generally good baseline for ensuring model accuracy. 

**Question 3: Next we are going to see how much our result in #2 depends on the fraction of data in
the training and test split. Repeat #2 but now every time through the loop split the data
so that only 10% is in the training set and 90% is in the test set. How does the
confidence interval for #3 compare to the confidence interval for #2?**

In [12]:
frac_correct_pred_estims = []
for trial in range(100):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.90)
    scaler = StandardScaler().fit(X_train)
    X_train = scaler.transform(X_train)
    X_test = scaler.transform(X_test)
    log_reg = LogisticRegression()
    log_reg.fit(X_train, y_train)
    frac_correct_pred_estims.append(np.mean(y_test == log_reg.predict(X_test)))
frac_correct_pred_estims.sort()
print(f"The 95% bootstrap confidence interval for the estimate fraction of correct predictions is [{frac_correct_pred_estims[1]}, {frac_correct_pred_estims[98]}]")

The 95% bootstrap confidence interval for the estimate fraction of correct predictions is [0.9220272904483431, 0.9727095516569201]


Our obtained 95% confidence interval of approximately (0.922, 0.973) demonstrates that a training-test split of 10/90% is slightly worse for model accuracy than the 75/25 split (from Task 2), which may indicate that we should allocate more of our data for training. 

**Question 4: Finally, we are going to test the importance of standardizing the data. Repeat #2 (so
back to 75% training and 25% test) but never standardize the X-variables. You will have
to make an additional change to the code. If not, for at least some of the times through
the loop you will get an error that the model did not converge (already showing that not
standardizing the X-variables is a problem). The default number of iterations for logistic
regression is 100, the code below shows how to set it to be 5,000 (which will be
sufficient for this problem, note: this will make the code take longer to run, the entire
loop of 100 should take 1 or 2 minutes):**

log_reg = LogisticRegression(max_iter=5000)

**How does the confidence interval for #4 compare to the confidence interval for #2?**

In [13]:
frac_correct_pred_estims = []
for trial in range(100):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)
    log_reg = LogisticRegression(max_iter=5000) #maximizes number of iterations for logreg model to prevent freezing
    log_reg.fit(X_train, y_train)
    frac_correct_pred_estims.append(np.mean(y_test == log_reg.predict(X_test)))
frac_correct_pred_estims.sort()
print(f"The 95% bootstrap confidence interval for the estimate fraction of correct predictions is [{frac_correct_pred_estims[1]}, {frac_correct_pred_estims[98]}]")

The 95% bootstrap confidence interval for the estimate fraction of correct predictions is [0.3146853146853147, 0.6923076923076923]


Our obtained 95% confidence interval of approximately (0.315, 0.692) demonstrates that standardizing the X-variables is very important for enhancing model accuracy because our observed accuracy confidence interval is far lower than that of Task 2.  